
# The initial file for the analysis with R
# is a text file "counts" (45,6Mo) generated by the featureCounts v1.6.2
# featureCounts is a program that counts how many reads map to genomic features
# https://rnnh.github.io/bioinfo-notebook/docs/featureCounts.html
# the program was applied to the bam files and generated de 55575 lines
# corresponding to the matching transcripts
# The first 7 columns are transcripts informations:
# "Geneid"    "Chr"       "Start"     "End"       "Strand"    "Length"    "gene_name"
# the other columns correspond to each bam file and their read numbers for each transcript
# The first line of the original count file contains the instructions used to generate the file:
# Program:featureCounts v1.6.2; Command:"featureCounts" "-a" "/media/DATA/bioinfo/genomes/mouse/UCSC_GRCh38_mm10/gencode.vM22.primary_assembly.annotation.gtf" "-o" "counts" "--extraAttributes" "gene_name" "-O" "-M" "-Q" "40" "-p" "-B" "-T" "24" "--tmpDir" "/tmp" 

# to import in R the file the following command was used:
# counts <- read.delim("counts", comment.char="#", stringsAsFactors=FALSE)

# it was then converted into a .rda file to reduce its size (max 25Mo in Github):
# save(counts, file = "counts.rda")
dir()
load("counts.rda")

get_meta_info = function(samples){
  samples = colnames(counts)[8:103]
  samples
  mouse = substring(samples,1,regexpr(samples, pattern = "_")-1)
  mouse
  rest = substring(samples,nchar(mouse)+2,nchar(samples))
  rest
  name_with_fod = regexpr(rest,pattern = "fod")==1
  name_with_fod
  rest[name_with_fod]=substring(rest[name_with_fod],nchar("fod")+2,nchar(rest[name_with_fod]))
  rest
  genotype = substring(rest,1,regexpr(rest, pattern = "_")-1)
  table(genotype)
  rest2 = substring(rest,nchar(genotype)+2,nchar(rest))
  rest2
  age = substring(rest2,1,regexpr(rest2, pattern = "_")-1)
  age
  rest3 = substring(rest2,nchar(age)+2,nchar(rest2))
  rest3
  
  table(age)
  age[age=="4mois"] = "4M"
  age[age=="8mois"] = "8M"
  age[age=="6sem"] = 6
  age[age=="8M"] = 8*4 #24
  age[age=="4M"] = 4*4 #24
  age[age=="6"] = "06"
  #levels(factor(age))
  #rest2
  
  extra_ = regexpr(rest3, pattern = "_")==1
  rest4 = rest3
  rest4[extra_] = substring(rest4[extra_],2,nchar(rest4[extra_]))
  rest4
  
  organ = (substring(rest4,1,nchar(rest4)-4))
  organ[organ == "Hippocampe"] = "hippocampe"
  organ = factor(organ)
  organ
  levels(organ) = c("Cerebellum", "Cortex","Hippocampus","Other")
  short_names_organs = factor(organ)
  levels(short_names_organs) =  c("Cer","Cor","Hip","Other")
  short_name_organ = as.character(short_names_organs)
  short_name = paste(short_name_organ, age, genotype, sep = "_")
  
  db =  data.frame(filename = samples, mouse = mouse,
                   name_with_fod, age=age, genotype=genotype,
                   organ=organ, short_name_organ=short_name_organ,
                   short_name=short_name, stringsAsFactors = TRUE)
  db
}

# Generation of the « count_matrix »  55573x96 of integers (number of reads)
# Name of each line is related to the transcrit :
# Example : ENSMUSG00000102693.1
# Name of the column correspond to short name of the sample names :
# Example : "Cer_06_WT"

sampleinfo = get_meta_info(colnames(counts)[8:103])
geneinfo = counts[,1:7]
counts_matrix = as.matrix(counts[8:103])
rownames(counts_matrix) = geneinfo$gene_name
colnames(counts_matrix) = sampleinfo$short_name
save(counts_matrix,file="count_matrix.rda")

# The table « sampleinfo » is generated from the name of the samples
# so that each sample has 8 columns with its metadata associated:
# "filename"         "mouse"            "name_with_fod"    "age"             "genotype"         "organ"            "short_name_organ" "short_name" 
# The column "mae_with_fod" has no particular interesting information
# 5 replicats for 2 génotypes for 3 structures for 3 ages
# an extra 6 samples from an other structure were used as controls
# 5*2*3*3 +6 = 96 samples
colnames(counts)
sampleinfo = get_meta_info(colnames(counts)[8:103])
save(sampleinfo,file = "sampleinfo.rda")


# Generation of a « geneinfo » table
# Keeping only the first 7 columns of the count table :
#   "Geneid"    "Chr"       "Start"     "End"       "Strand"    "Length"    "gene_name"

geneinfo = counts[,1:7]
save(geneinfo, file = "geneinfo.rda")
